### PR TITLE
[WIP][PipelineRL] Normalization of new_logprobs and addition of other RL metrics

### PR DESCRIPTION
**Pull request by bigximik**
Add generic denominator_batch_field to LossDef so any metric can be normalized by a pre-computed per-micro-batch scalar from the batch context, bypassing TP/SP/PP splitting entirely.

For grpo_new_logprobs: compute num_docs = (labels_per_document > 0).sum() in language_model.py before any parallel splitting, giving a true per-document average regardless of variable document lengths.

Only sequence_data_rank==0 contributes to num_docs. The runner all_reduces the denominator across the data group (which includes SDP ranks); if every SDP rank reported its own num_docs, a single document processed by SDP=2 would be counted twice, halving the metric.

Also clamp num_labels_in_seq to avoid 0/0=nan for padding segments or fully-masked documents (loss_mask=0 there so the numerator is 0 too).

Tests verify:
- num_docs counts only unmasked documents
- padding segments (pad_to_size) are excluded
- with SDP=2, only rank 0 contributes num_docs so the all_reduce SUM across SDP ranks gives the correct denominator

# ✨ Description

Please provide a brief summary of the changes, relevant motivation, and context.
Include any related issue numbers or links to discussions, and explain why this change is necessary.

Closes # 

## 🔍 Type of change

Select all that apply:

- [ ] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [ ] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

## 📝 Changes

List the key changes introduced in this PR:

1. Change A
2. Change B

## ✅ Checklist

Make sure the following tasks are completed before submitting the PR:

### General

- [ ] 📜 I have read and followed the [contributing guidelines](https://servicenow.github.io/Fast-LLM/contributing/contributing).
- [ ] 🏷️ I am using a clear and descriptive PR title that summarizes the key change or feature introduced.
- [ ] 🎉 The functionality is complete, and I have tested the changes.
- [ ] 📝 I have updated the documentation if needed.
- [ ] ⚠️ The change does not introduce any new issues (e.g., runtime warnings, type checker errors, linting problems, unhandled edge cases).
- [ ] 🧩 I have commented my code, especially in hard-to-understand areas.

### Dependencies and Configuration

- [ ] 🐋 I have updated the Docker configuration or dependencies, if applicable.
- [ ] 🔄 I have ensured compatibility with the existing setup after dependency changes.

### Testing

- [ ] 🧪 I have added or updated tests to cover my changes.
- [ ] ✔️ New and existing tests pass locally with my changes.
- [ ] 🚦 I have tested these changes on GPUs and verified training stability.
- [ ] 🏋️ I have tested the changes on realistic training workloads, if applicable.

### Performance Impact

- [ ] 📊 I have run benchmarks where applicable to evaluate the performance impact.
- [ ] ✅ The benchmarks show no performance regression.
- [ ] 🚀 The benchmarks indicate a potential performance improvement.
- [ ] ⚠️ The benchmarks indicate a potential performance degradation.
- [ ] 📈 I have provided benchmark results and detailed any performance impact below, if applicable.

## 📊 Performance Impact Details

If there is any impact on performance, describe it and provide benchmark results, if applicable:

---

## 🗒️ Additional Notes

Include any additional context, information, or considerations here, such as known issues, follow-up tasks, or backward compatibility concerns.

### CODE REVIEW

File: fast_llm/data/document/language_model.py [hunk 1]
```diff
@@ -39,6 +39,9 @@ class LanguageModelTargetInput(ModelInput):
 class LanguageModelInput(TokenModelInput):
     targets: list[LanguageModelTargetInput] = dataclasses.field(default_factory=list)
     image_patches: PatchModelInput | None = None
+    # Number of documents with at least one response token in this micro-batch.
```
Comment (jlamypoirier) [inline fast_llm/data/document/language_model.py:42]: Why not just using the plain document count? Is there any scenario where we would expect response-free documents and need to take them into account? Seems like they shouldn't be in the batch to begin with because they don't contribute to the loss...

File: fast_llm/engine/base_model/config.py [hunk 2]
```diff
@@ -109,6 +109,9 @@ class LossDef:
     name: str
     formatted_name: str
     # The number of times this loss is evaluated by the model for each micro-batch. Used as a denominator for averaging.
-    # TODO: Allow variable count?  Would need a reduction across PP devices.
     count: int = 1
     dtype: DataType = DataType.float32
+    # If set, normalize this metric by summing values from context.batch[i][denominator_batch_field]
+    # across micro-batches and DP ranks, instead of using count * data_parallel * num_inputs.
+    # The field must be a scalar (int or float) pre-computed before any TP/SP/PP splitting.
+    denominator_batch_field: str | None = None
```
Comment (jlamypoirier) [inline fast_llm/engine/base_model/config.py:117]: Somewhat redundant with `count`. Forcing the denominator to be in a scalar-valued kwarg field may not be generic enough, I'd suggest returning losses as a `(numerator, denominator)` tuple instead, or maybe reducing the denominator beforehand?

File: fast_llm/data/document/language_model.py [hunk 3]
```diff
@@ -161,6 +171,21 @@ def _get_model_input(
...
+                # normalization. Only counted on sequence_data_rank==0 so that when the runner
+                # all_reduces the denominator across the data group (which includes SDP ranks),
+                # each document is counted exactly once even though all SDP ranks see the same
+                # documents (but process different token slices of them).
+                if model_input.num_docs is None:
+                    # Skip .item() on meta tensors (called during setup/shape inference).
+                    if labels_per_document.is_meta:
+                        model_input.num_docs = 0
+                    else:
+                        model_input.num_docs = (
```
Comment (jlamypoirier) [inline fast_llm/data/document/language_model.py:184]: We could just reduce over the batch data group instead?

File: fast_llm/engine/schedule/runner.py [hunk 4]
```diff
@@ -287,20 +287,39 @@ def run_step(
...
                 if losses:
-                    loss_count = (
-                        self._loss_definitions[name].count
-                        * self._distributed_config.data_parallel
-                        * context.schedule.config.num_inputs
-                    )
-                    reduced_loss = torch.stack(losses).sum() / loss_count
-                    if self._distributed.data_group:
-                        all_reduce(reduced_loss, group=self._distributed.data_group)
+                    denom_field = loss_def.denominator_batch_field
```
Comment (jlamypoirier) [inline fast_llm/engine/schedule/runner.py:294]: This works for metrics, but not for actual losses used for training which would need to reproduce this code. I think this could be addressed by reducing the values at the beginning of the batch, ex. in `ModelInput.reduce_counts` so we can use the counts freely from anywhere.

File: tests/data/test_preprocessing.py [hunk 5]
```diff
@@ -53,3 +55,163 @@ def test_preprocessing(tokens, loss_masking_spans):
 
     Assert.eq(len(model_input.targets), 1)
     Assert.all_equal(model_input.targets[0].tokens, torch.cat(label_tokens)[1:])
+
```
Comment (jlamypoirier) [inline tests/data/test_preprocessing.py:58]: I think most of these can be combined into a single parameterized test?

**Review by jlamypoirier — COMMENTED** [review]
Looks good, but I would like to look at the problem in a more generic context, especially since we've had multiple requests to address normalization for all losses and metrics (ex. #426).

File: fast_llm/data/document/language_model.py (hunk 1 above)
Comment (bigximik) [reply fast_llm/data/document/language_model.py:42]: Agree that response-free documents shouldn't be in the batch. However the > 0 check is needed because of padding — by the time we reach _get_model_input, padding is just another entry in lengths and structurally indistinguishable from a real document. To get the true document count cleanly, we can store num_real_docs = len(documents) in from_documents before
   padding is appended.

File: fast_llm/engine/base_model/config.py (hunk 2 above)
Comment (bigximik) [reply fast_llm/engine/base_model/config.py:117]: The reason of not returning (numerator, denominator) from the loss is that the loss is computed per micro-batch split (micro sequence), and a document can span a split boundary. If each split returned its own document count as the denominator, documents straddling a boundary would be counted twice. That's why num_docs is computed only on micro_sequence_index == 0 and
  all_reduced — to count each document exactly once.

The current design handles this by precomputing num_docs at the batch level before splitting, and having the loss reference it by name via denominator_batch_field — essentially "use this precomputed scalar as my denominator" rather than computing it from within the split. Returning a denominator from the loss function directly would only work cleanly for
  quantities that are local to a split (like num_tokens or num_labels), but num_docs inherently crosses split boundaries so it can't be accumulated that way easily.

Comment (jlamypoirier) [reply fast_llm/engine/base_model/config.py:117]: In either case what matters is `sum(numerator)/sum(denominator)`, so as long as the returned denominators sum to the right value (as in the valuers you're using) things still work

Anyway in #477 I'm precomputing `sum(denominator)` so we simply return `numerator/sum(denominator)` in losses and the reduction is a simple sum.

**Comment by bigximik** [comment]
superseded by other PRs
